### PR TITLE
[AArch64] Remove redundant instructions in int-to-fp of lowest vector…

### PR DESCRIPTION
**Pull request by SpencerAbson**
… element. 
When converting the lowest element (that in lane 0) of a vector from an integer to a floating-point value, LLVM should select the SIMD scalar variant of CVTF (https://developer.arm.com/documentation/dui0801/g/A64-SIMD-Scalar-Instructions/SCVTF--scalar--integer-) to avoid the FPR to GPR register transfers that are required to use the general floating-point variant (https://developer.arm.com/documentation/dui0801/g/A64-Floating-point-Instructions/SCVTF--scalar--integer-). This is possible as the lowest element can be referred to by the corresponding scalar sub-register with the width of the vector's constituent elements.

This patch adds new TableGen patterns to remove these redundant instructions for AArch64, as well as back-end tests to ensure the new preferred instruction selection result is produced. Existing tests that relied on the previous selection result have also been updated.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -6068,6 +6068,32 @@ def : Pat<(f16 (any_sint_to_fp (i32 (any_fp_to_sint f16:$Rn)))),
 def : Pat<(f16 (any_uint_to_fp (i32 (any_fp_to_uint f16:$Rn)))),
           (UCVTFv1i16 (f16 (FCVTZUv1f16 f16:$Rn)))>;
 }
+
+// int -> float conversion of value in lane 0 of simd vector should use 
+// correct cvtf variant to avoid costly fpr <-> gpr register transfers.
+def : Pat<(f32 (sint_to_fp (i32 (vector_extract (v4i32 FPR128:$Rn), (i64 0))))),
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:6074]: Do these need a HasNEONandIsStreamingSafe predicate?

Comment (SpencerAbson) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td:6074]: I believe so - the entire section that these patterns are in is predicated by a HasNeonandIsStreamingSafe slightly further up, though admittedly this is not that clear.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 2]
```diff
@@ -6068,6 +6068,32 @@ def : Pat<(f16 (any_sint_to_fp (i32 (any_fp_to_sint f16:$Rn)))),
...
+def : Pat<(f64 (sint_to_fp (i64 (vector_extract (v2i64 FPR128:$Rn), (i64 0))))),
+          (SCVTFv1i64 (i64 (EXTRACT_SUBREG (v2i64 FPR128:$Rn), dsub)))>;
+
+def : Pat<(f64 (uint_to_fp (i64 (vector_extract (v2i64 FPR128:$Rn), (i64 0))))),
+          (UCVTFv1i64 (i64 (EXTRACT_SUBREG (v2i64 FPR128:$Rn), dsub)))>;
+
+// fp16: integer extraction from vector must be at least 32-bits to be legal.
+// Actual extraction result is then an in-reg sign-extension of lower 16-bits.
+let Predicates = [HasNEONandIsStreamingSafe, HasFullFP16] in {
+def : Pat<(f16 (sint_to_fp (i32 (sext_inreg (i32 (vector_extract (v8i16 FPR128:$Rn), (i64 0))), i16)))),
```
Comment (david-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:6089]: nit: Can you format both of these patterns so they meet the 80 character line limit? Thanks!

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 1 above)
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td:6074]: I agree with @davemgreen. It looks like these patterns need guarding with `let Predicates = [HasNEONandIsStreamingSafe]` similar to the two fp16 patterns below.

Comment (SpencerAbson) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td:6074]: Apologies if my comment was unclear, these patterns already fall in the scope of a previous `let Predicates = [HasNEONandIsStreamingSafe]` binding. I can always another guard above these patterns if we'd like this to be clearer?

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 3]
```diff
@@ -6068,6 +6068,34 @@ def : Pat<(f16 (any_sint_to_fp (i32 (any_fp_to_sint f16:$Rn)))),
...
+
+def : Pat<(f64 (uint_to_fp (i64 (vector_extract (v2i64 FPR128:$Rn), (i64 0))))),
+          (UCVTFv1i64 (i64 (EXTRACT_SUBREG (v2i64 FPR128:$Rn), dsub)))>;
+
+// fp16: integer extraction from vector must be at least 32-bits to be legal.
+// Actual extraction result is then an in-reg sign-extension of lower 16-bits.
+let Predicates = [HasNEONandIsStreamingSafe, HasFullFP16] in {
+def : Pat<(f16 (sint_to_fp (i32 (sext_inreg (i32 (vector_extract 
+          (v8i16 FPR128:$Rn), (i64 0))), i16)))), 
+              (SCVTFv1i16 (f16 (EXTRACT_SUBREG (v8i16 FPR128:$Rn), hsub)))>;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:6091]: I'm not sure if 80 columns in this file ever really helps makes things cleaner. The "input" and the "output" are good to line up, to make the patterns easier to read:
```
def : Pat<(f16 (sint_to_fp (i32 (sext_inreg (i32 (vector_extract 
                  (v8i16 FPR128:$Rn), (i64 0))), i16)))), 
          (SCVTFv1i16 (f16 (EXTRACT_SUBREG (v8i16 FPR128:$Rn), hsub)))>;
```

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td:6074]: Ah I see. I mis-read where the above predicate got closed.

**Review by davemgreen — APPROVED** [review]
LGTM, thanks.

**Pull request merged**